ggml/src/ggml-metal.m

```diff
@@ -3548,11 +3553,154 @@ void ggml_backend_metal_capture_next_compute(ggml_backend_t backend) {
     ctx->capture_next_compute = true;
 }
 
-ggml_backend_t ggml_backend_reg_metal_init(const char * params, void * user_data); // silence warning
+// backend device
+
+static const char * ggml_backend_metal_device_get_name(ggml_backend_dev_t dev) {
+    return "Metal";
+
+    GGML_UNUSED(dev);
+}
 
-ggml_backend_t ggml_backend_reg_metal_init(const char * params, void * user_data) {
+static const char * ggml_backend_metal_device_get_description(ggml_backend_dev_t dev) {
+    return [[g_state.mtl_device name] UTF8String];
```

Comment:
I don't think there is a guarantee that `mtl_device` is initialized here, it probably needs a call to `ggml_backend_metal_get_device`/`ggml_backend_metal_free_device` like in `ggml_backend_metal_device_get_memory`. However, I imagine that could cause issues with the lifetime of the string returned by `MTLDevice`, so it may be necessary to keep a copy of the string in the context instead.

Reply:
Should be fixed now.

I also reworked the implementation to avoid accessing `g_state` when we can get the device context locally. Should be much cleaner now and easier to add multi-GPU support in the future if needed.